### PR TITLE
[MRG] New feature : Custom metrics for dtw_path (metrics.dtw_path_from_metric)

### PR DESCRIPTION
**Pull request by rfayat**
Following the discussion about adding the possibility to use different distance metrics for computing the dynamical time wrapping #162 "Haversine distance for the metrics modules", I added **dtw_path_from_metric** to tslearn.metrics.

It extends @rtavenar's suggestion to create a new function (rather than modifying dtw_path) taking as input a precomputed distance matrix between each pair of points of two time series to wrap by also allowing to give as input a string referring to a function compatible with scipy's [cdist](https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.distance.cdist.html) or directly by passing a callable that will be used for the computation of pairwise distances (similar to the syntax of sklearn [pairwise_distances](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.pairwise_distances.html)).

Similarly to dtw_path, the function returns both the path and a similarity metric which is the sum of the input distance metric for the two wrapped time series.

For instance, to compute the path in the same way as dtw_path, the squared euclidean distance can be used equivalently in three ways:
- By precomputing the distance matrix :
```python
from scipy.spatial.distance import cdist
precomputed_distance = cdist(s1, s2, metric="sqeuclidean")
path, similarity = dtw_path_from_metric(dist_matrix, metric="precomputed")
```
- By specifying a metric compatible with cdist:
```python
path, similarity = dtw_path_from_metric(s1, s2, metric="sqeuclidean")
```
- By creating a custom function computing the squared euclidean distance:
```python
import numpy as np
def sqeuclidean(x, y):
    return(np.sum((x-y)**2))
path, similarity = dtw_path_from_metric(s1, s2, metric=sqeuclidean)
```
In this case, the output similarity would be the square of the one returned by dtw_path (sum of squared euclidean distances for each point vs euclidean distance).

NDLR: Thank you for this great toolbox ! This is my first pull request so I am 100% open to feedback / suggestions.

Cheers,

Romain

### CODE REVIEW

File: tslearn/metrics.py [hunk 1]
```diff
@@ -21,6 +22,17 @@
 GLOBAL_CONSTRAINT_CODE = {None: 0, "": 0, "itakura": 1, "sakoe_chiba": 2}
 VARIABLE_LENGTH_METRICS = ["dtw", "gak", "softdtw"]
 
+_VALID_METRICS = ['euclidean', 'braycurtis', 'canberra', 'chebyshev',
```
Comment (rtavenar) [inline tslearn/metrics.py:25]: Why not import those from scikit-learn? They will probably be more reactive than we can be if new metrics are added in scipy, I guess.

Comment (rfayat) [reply tslearn/metrics.py:25]: Agreed, I'll even replace scipy's cdist by scikit's pairwise_distances and let it take care directly of what metrics are allowed or not.

**Comment by rtavenar** [comment]
Hi @rfayat ,

Thanks, this is now very close to being ready for merge. I will leave final comments in a minute.

File: tslearn/docs/dtw.rst [hunk 2]
```diff
@@ -85,6 +85,30 @@ simplicity):
        return sqrt(C[n, m])
 
 
+Using a different ground metric
```
Comment (rtavenar) [inline tslearn/docs/dtw.rst:88]: I took the freedom to change this page so that the paragraph is included earlier, which looks more natural, I think.

File: tslearn/docs/examples/plot_dtw_custom_metric.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+# Definitions for the axes
+left, bottom = [.01, .51], .1
+w_ts, h_ts = .1, .2
+left_h = [l + w_ts + .01 for l in left]
+width, height = .325, .65
+bottom_h = bottom + height + .02
+
+# Create the axes and plot the distance matrices
+plt.figure(1, figsize=(12, 6))
```
Comment (rtavenar) [inline tslearn/docs/examples/plot_dtw_custom_metric.py:101]: This example is great. I have maybe one minor remark: it would be better to generate two distinct figures and pick the second one as the thumbnail in the gallery, as done there: https://github.com/rtavenar/tslearn/blob/master/tslearn/docs/examples/metrics/plot_lb_keogh.py

File: tslearn/metrics.py [hunk 4]
```diff
@@ -207,6 +209,188 @@ def dtw_path(s1, s2, global_constraint=None, sakoe_chiba_radius=None,
...
+
+    Valid values for metric are the same as for scikit-learn
+    `pairwise_distances`_ function I.E. a string (E.G. "euclidean",
+    "sqeuclidean", "hamming") or a function that is used to compute the
+    pairwise distances. See `scikit`_ and `scipy`_ documentations for more
+    information about the available metrics.
+
+    Parameters
+    ----------
+    s1 : array, shape = (sz1, sz2) if metric=="precomputed", (sz1,) otherwise
```
Comment (rtavenar) [inline tslearn/metrics.py:266]: I'd rather set the `(sz1, d)` case as default than the precomputed case. (also add the feature dimension d for s2)

File: tslearn/metrics.py [hunk 5]
```diff
@@ -207,6 +209,188 @@ def dtw_path(s1, s2, global_constraint=None, sakoe_chiba_radius=None,
...
+    the user and return both the path and the similarity.
+
+    Similarity is computed as the cumulative cost along the wrapped time
+    series.
+
+    It is not required that both time series share the same size, but they must
+    be the same dimension. DTW was originally presented in [1]_.
+
+    Valid values for metric are the same as for scikit-learn
+    `pairwise_distances`_ function I.E. a string (E.G. "euclidean",
```
Comment (rtavenar) [inline tslearn/metrics.py:259]: I would not use capital letters for i.e. nor e.g.

**Comment by rtavenar** [comment]
And it would be good to fix PEP8 issues too.

File: tslearn/docs/dtw.rst (hunk 2 above)
Comment (rfayat) [reply tslearn/docs/dtw.rst:88]: Looks better indeed !

File: tslearn/metrics.py (hunk 4 above)
Comment (rfayat) [reply tslearn/metrics.py:266]: Works for me, I'll also change the default metric to "euclidean" instead of "precomputed" to be consistent with the default input.

**Comment by rfayat** [comment]
Hi @rtavenar,

Thank you for your feedback, I've just added the requested changes. I still could'nt solve all PEP8 issue (the two remaining are due to urls in metrics.dtw_path_from_metric's doc making the corresponding lines longer than 79 caracters), hope this won't be an issue.

Cheers

**Comment by rtavenar** [comment]
Thanks a lot @rfayat !

This is really good work.
Since your PR points to `master`, I'll probably wait a bit before merging (I'd like to merge some of the currently open PRs to `dev`, then merge `dev` to `master`, branches are a bit messy at the moment), but it will be integrated in tslearn v0.4 for sure.
